File: university-database/src/main/java/by/arhor/university/database/model/Module.java
package by.arhor.university.database.model;

import java.io.Serializable;
import java.util.Objects;
import java.util.StringJoiner;

import javax.xml.bind.annotation.XmlAccessType;
import javax.xml.bind.annotation.XmlAccessorType;
import javax.xml.bind.annotation.XmlAttribute;
import javax.xml.bind.annotation.XmlElement;
import javax.xml.bind.annotation.XmlRootElement;
import javax.xml.bind.annotation.XmlTransient;

@XmlRootElement(name = "module")
@XmlAccessorType(XmlAccessType.FIELD)
public final class Module implements Serializable {

  @XmlAttribute
  private String name;

  @XmlElement(name = "dependencies")
  private Dependencies dependencies;

  @XmlElement(name = "queries")
  private Queries queries;

  @XmlTransient
  private boolean resolved;

  @XmlTransient
  private int executionCost;

  public String getName() {
    return name;
  }

  public void setName(String name) {
    this.name = name;
  }

  public Dependencies getDependencies() {
    return dependencies;
  }

  public void setDependencies(Dependencies dependencies) {
    this.dependencies = dependencies;
  }

  public Queries getQueries() {
    return queries;
  }

  public void setQueries(Queries queries) {
    this.queries = queries;
  }

  public boolean isResolved() {
    return resolved;
  }

  public void setResolved(boolean resolved) {
    this.resolved = resolved;
  }

  public int getExecutionCost() {
    return executionCost;
  }

  public void setExecutionCost(int executionCost) {
    this.executionCost = executionCost;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    Module module = (Module) o;
    return Objects.equals(name, module.name)
        && Objects.equals(dependencies, module.dependencies)
        && Objects.equals(queries, module.queries);
  }

  @Override
  public int hashCode() {
    return Objects.hash(name, dependencies, queries);
  }

  @Override
  public String toString() {
    return new StringJoiner(", ", Module.class.getSimpleName() + "[", "]")
        .add("name='" + name + "'")
        .add("dependencies=" + dependencies)
        .add("queries=" + queries)
        .toString();
  }
}

File: university-database/src/main/scala/by/arhor/university/database/model/Module.scala
package by.arhor.university.database.model

import java.io.Serializable
import java.util.function.Consumer

import javax.xml.bind.annotation._

@xmlRootElement(name = "module")
@xmlAccessorType(XmlAccessType.FIELD)
case class Module(
    @xmlAttribute var name: String,
    @xmlElement   var dependencies: Dependencies,
    @xmlElement   var queries: Queries,
    @xmlTransient var resolved: Boolean,
    @xmlTransient var executionCost: Int
) extends Serializable with Comparable[Module] {
  def this() = this(null, null, null, false, 0)

  override def compareTo(that: Module): Int = this.executionCost - that.executionCost

  def forEachDependency(action: Consumer[Dependency]): Unit = {
    if (dependencies != null) dependencies.forEach(action)
  }

  def forEachQuery(action: Consumer[Query]): Unit = {
    if (queries != null) queries.forEach(action)
  }
}
